### PR TITLE
use max-heap to collect shink candidates by byte savings

### PR DESCRIPTION
**Pull request by HaoranYi**
#### Problem

1. selecting storage to shink by alive ratio may not improve the alive ratio as much as by byte savings. 

For example, given two storages below:
Storage A: alive = 1, total = 2,    alive ratio = 0.5
Storage B: alive = 6, total = 10,  alive ratio = 0.6

If we select by ratio, then A will be chosen to shrink, which gives result ratio of 7/11 (0.63).
However, if we select by bytes savings, B will be chosen. This gives the result ratio of 7/8 (0.875).

2. When collecting candidate storages for shrinking, we don't need to sort all the
storages. We could just partially sort the storages before the cutoff (i.e. use a heap).

#### Summary of Changes

Use max-heap to store and partially sort and select candidate storage for shrinking by bye savings.

Fixes #

### CODE REVIEW

**Comment by HaoranYi** [comment]
> How are the performance numbers, before and after?

image
looks like there is not much performance difference (red: master (#33739), blue: this pr).

File: accounts-db/src/accounts_db.rs [hunk 1]
```diff
@@ -4322,33 +4340,33 @@ impl AccountsDb {
...
                 continue;
             };
-            candidates_count += 1;
             total_alive_bytes += Self::page_align(store.alive_bytes() as u64);
             total_bytes += store.capacity();
             let alive_ratio =
                 Self::page_align(store.alive_bytes() as u64) as f64 / store.capacity() as f64;
+
+            let current_store_size = store.capacity();
+            let after_shrink_size = Self::page_align(store.alive_bytes() as u64);
```
Comment (HaoranYi) [inline accounts-db/src/accounts_db.rs:4349]: Do we still need `page_align`? I remember we made a change recently that we no longer require that append_vec to be page_align?

Comment (jeffwashington) [reply accounts-db/src/accounts_db.rs:4349]: it still needs to be page align. It just isn't padded to a page boundary.

File: accounts-db/src/accounts_db.rs [hunk 2]
```diff
@@ -4304,13 +4304,31 @@ impl AccountsDb {
...
+            fn eq(&self, other: &Self) -> bool {
+                self.bytes_saved == other.bytes_saved
+            }
+        }
+        impl Eq for StoreUsageInfo {}
+
+        // create a max-heap to store the StorageUsageInfo by saved bytes
+        let mut store_usage: std::collections::BinaryHeap<StoreUsageInfo> =
+            std::collections::BinaryHeap::new();
+        store_usage.reserve(shrink_slots.len());
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: yeah.

File: accounts-db/src/accounts_db.rs [hunk 3]
```diff
@@ -2442,6 +2442,11 @@ pub struct PubkeyHashAccount {
     pub account: AccountSharedData,
 }
 
+struct SelectedShrinkCandidates {
+    pub to_shrink: HashMap<Slot, Arc<AccountStorageEntry>>,
+    pub next_batch: ShrinkCandidates,
```
[the remark this answers is not in the document]
Comment (HaoranYi) [reply accounts-db/src/accounts_db.rs]: no. done.

**Comment by HaoranYi** [comment]
> How do the byte savings look on mnb and/or pop-net?

image

comparison of mnb run from last night shows not much difference.

blue: this pr
red: master

**Pull request closed without merging**